datafusion/optimizer/src/type_coercion.rs

```diff
@@ -816,6 +830,55 @@ mod test {
         Ok(())
     }
 
+    #[test]
+    fn concat_for_type_coercion() -> Result<()> {
+        use datafusion_expr::BuiltinScalarFunction::{Concat, ConcatWithSeparator};
+
+        let empty = empty_with_type(DataType::Utf8);
+        let args = vec![col("a"), lit("b"), lit(true), lit(false), lit(13)];
+
+        // concat
+        {
+            let expr = Expr::ScalarFunction {
+                fun: Concat,
+                args: args.clone(),
+            };
```

Comment:
I think you can use the `concat` builder function here https://docs.rs/datafusion/12.0.0/datafusion/logical_plan/fn.concat.html

```suggestion
            let expr = concat(args.clone());
```